http/middleware.go

```diff
@@ -0,0 +1,102 @@
+package http
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"net/http"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+func HTTPLoggingMW(logger *zap.Logger) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			srw := &statusResponseWriter{
+				ResponseWriter: w,
+			}
+
+			var buf bytes.Buffer
+			r.Body = &bodyEchoer{
+				rc:    r.Body,
+				teedR: io.TeeReader(r.Body, &buf),
+			}
+
+			defer func(start time.Time) {
+				errField := zap.Skip()
+				if errStr := w.Header().Get(PlatformErrorCodeHeader); errStr != "" {
+					errField = zap.Error(errors.New(errStr))
+				}
+
+				errReferenceField := zap.Skip()
+				if errReference := w.Header().Get(PlatformErrorCodeHeader); errReference != "" {
+					errReferenceField = zap.String("error_code", PlatformErrorCodeHeader)
+				}
+
+				fields := []zap.Field{
+					zap.String("method", r.Method),
+					zap.String("host", r.Host),
+					zap.String("path", r.URL.Path),
+					zap.String("query", r.URL.Query().Encode()),
+					zap.String("proto", r.Proto),
+					zap.Int("status_code", srw.code()),
+					zap.Int("response_size", srw.responseBytes),
+					zap.Int64("content_length", r.ContentLength),
+					zap.String("referrer", r.Referer()),
+					zap.String("remote", r.RemoteAddr),
+					zap.String("user_agent", r.UserAgent()),
+					zap.Duration("took", time.Since(start)),
+					errField,
+					errReferenceField,
+				}
+
+				isValidMethodFn, ok := blacklistEndpoints[r.URL.Path]
+				if !ok || isValidMethodFn(r.Method) {
+					fields = append(fields, zap.ByteString("body", buf.Bytes()))
+				}
+
+				logger.Debug("Request", fields...)
+			}(time.Now())
+
+			next.ServeHTTP(srw, r)
+		}
+		return http.HandlerFunc(fn)
+	}
+}
+
+type isValidMethodFn func(method string) bool
+
+var (
+	ignoreMethod = func(string) bool { return true }
+
+	blacklistEndpoints = map[string]isValidMethodFn{
```

Comment:
Nice pattern 👍

Can we also add write and query endpoints to the blacklist? They get hit a lot, and we really don't need to log all those requests.

Reply:
great point, I'll definitely add that in 👍

Reply:
@jademcgough pushed up the changes 👍

Reply:
@jademcgough I went through more of the request and found a few more that would have secrets. The org endpoints I totally missed.

Reply:
@jsteenb2 ah right, I was pretty sure there was something I was forgetting yesterday, it was the org secrets path :)